File: tests/fixtures/pii.py
"""Gera PII válida em formato, no momento do teste.

**Nenhum valor daqui é literal no repositório.** Eles nascem em memória e morrem no fim
da execução — é o que permite que a varredura do portão de segurança seja absoluta, sem
lista de exceções (CLAUDE.md 13b, docs/planos/00-fixtures-pii.md).

As duas alternativas que isto substitui:

- literal sintético mais lista de exceções na varredura → lista de exceções em varredura
  de segurança é a mesma porta que se recusa a abrir no guardrail;
- valores fora do formato real → enfraquece exatamente o teste que se quer fazer.

Ganho que não estava no problema: um literal exercita **um** formato de CPF; um gerador
semeado exercita cem, e é onde aparece o zero à esquerda e a variação que um exemplo
escolhido a dedo esconde.
"""

from __future__ import annotations

import random

_LETRAS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
_DDDS = ("11", "21", "31", "41", "47", "48", "51", "61", "62", "71", "81", "85")
#: Domínios reservados pela RFC 2606 — não são registráveis por ninguém.
_DOMINIOS = ("example.com", "example.org", "example.net")


def _cpf(rng: random.Random) -> str:
    """Com dígitos verificadores corretos: formato perfeito, conteúdo inventado."""
    n = [rng.randint(0, 9) for _ in range(9)]
    for _ in range(2):
        s = sum((len(n) + 1 - i) * v for i, v in enumerate(n))
        d = (s * 10) % 11
        n.append(0 if d == 10 else d)
    return f"{n[0]}{n[1]}{n[2]}.{n[3]}{n[4]}{n[5]}.{n[6]}{n[7]}{n[8]}-{n[9]}{n[10]}"


def _cep(rng: random.Random) -> str:
    # Inclui prefixos com zero à esquerda de propósito: é a variação que quebra
    # implementações ingênuas e que o gerador expõe ao longo de vários seeds.
    return f"{rng.randint(1000, 99999):05d}-{rng.randint(0, 999):03d}"


def _telefone(rng: random.Random) -> str:
    return f"+55 {rng.choice(_DDDS)} 9{rng.randint(1000, 9999)}-{rng.randint(1000, 9999)}"


def _placa(rng: random.Random) -> str:
    """Padrão Mercosul: LLLNLNN."""
    letras = "".join(rng.choice(_LETRAS) for _ in range(3))
    return f"{letras}{rng.randint(0, 9)}{rng.choice(_LETRAS)}{rng.randint(0, 9)}{rng.randint(0, 9)}"


def _email(rng: random.Random) -> str:
    nome = "".join(rng.choice("abcdefghijklmnopqrstuvwxyz") for _ in range(rng.randint(4, 9)))
    sep = rng.choice([".", "_", ""])
    sobre = "".join(rng.choice("abcdefghijklmnopqrstuvwxyz") for _ in range(rng.randint(4, 9)))
    return f"{nome}{sep}{sobre}@{rng.choice(_DOMINIOS)}"


def gerar_pii(seed: int) -> dict[str, str]:
    """Semeado ⇒ determinístico. O mesmo seed dá o mesmo conjunto, sempre.

    O seed é sempre explícito no teste, nunca `random.seed()` global — senão a ordem
    de execução muda o valor e a falha deixa de ser reproduzível.
    """
    rng = random.Random(seed)
    return {
        "cpf": _cpf(rng),
        "cep": _cep(rng),
        "email": _email(rng),
        "telefone": _telefone(rng),
        "placa": _placa(rng),
    }


def frase_do_lead(seed: int) -> tuple[str, dict[str, str]]:
    """A fala e os valores, para o teste assertar sobre os dois.

    O formato imita o do dataset do desafio: PII solta no meio de texto livre, em
    ordem variada, misturada com o dado de qualificação (a idade) que **não** é PII e
    não pode ser mascarado — cotar depende dele.
    """
    p = gerar_pii(seed)
    return (
        f"Tenho 35 anos, CPF {p['cpf']}, CEP {p['cep']}, "
        f"meu email é {p['email']} e o whats é {p['telefone']}, "
        f"placa {p['placa']}"
    ), p


def cep_de(prefixo: str, *, com_hifen: bool = True) -> str:
    """Constrói um CEP a partir do prefixo, em vez de escrever o literal.

    O que os testes de preço e de replay exercitam é o **prefixo** — os dois dígitos
    que disparam o agravo de 1,30 (`07`, `08`, `21`, `26`, `59`). O resto é
    preenchimento.

    Construir em vez de literalizar mantém a varredura de PII absoluta, sem exceção
    por arquivo (CLAUDE.md 13a/13b), e deixa mais claro o que a linha testa: quem lê
    `cep_de("07")` entende o caso na hora; quem lê `"07145-200"` precisa contar
    dígitos para saber qual é o ponto.
    """
    corpo = f"{prefixo}000"
    return f"{corpo}-000" if com_hifen else f"{corpo}000"

File: tests/web/fixtures/pii.ts
/**
 * Gera PII válida em formato, no momento do teste — a porta do TypeScript para a
 * regra do invariante 13b do `CLAUDE.md`.
 *
 * **Nenhum valor daqui é literal no repositório.** Eles nascem em memória e
 * morrem no fim da execução, e é o que permite que a varredura do portão de
 * segurança seja absoluta, **sem lista de exceções** — que é a mesma porta que
 * este projeto se recusa a abrir no guardrail.
 *
 * O gerador é semeado com um **LCG explícito**: `Math.random()` não aceita seed,
 * e sem seed uma falha deixa de ser reproduzível. Mesmo desenho e mesmo contrato
 * `(frase, pii)` do gerador em Python.
 *
 * Nenhum golden file com a saída: gravar o resultado traria o literal de volta
 * pela porta dos fundos.
 */

const LETRAS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const MINUSCULAS = "abcdefghijklmnopqrstuvwxyz";
const DDDS = ["11", "21", "31", "41", "47", "48", "51", "61", "62", "71", "81", "85"];
/** Domínios reservados pela RFC 2606 — não são registráveis por ninguém. */
const DOMINIOS = ["example.com", "example.org", "example.net"];
/** Montado em runtime: o literal do prefixo internacional não fica no fonte. */
const CODIGO_PAIS = 55;

/** Linear congruential generator — o mesmo `seed` dá a mesma sequência, sempre. */
function lcg(seed: number) {
  let estado = (seed >>> 0) || 1;
  return {
    /** Inteiro em [min, max]. */
    inteiro(min: number, max: number): number {
      estado = (Math.imul(estado, 1664525) + 1013904223) >>> 0;
      return min + (estado % (max - min + 1));
    },
    escolher<T>(itens: readonly T[]): T {
      estado = (Math.imul(estado, 1664525) + 1013904223) >>> 0;
      const item = itens[estado % itens.length];
      if (item === undefined) throw new Error("lista vazia");
      return item;
    },
  };
}

type Rng = ReturnType<typeof lcg>;

/** Com dígitos verificadores corretos: formato perfeito, conteúdo inventado. */
function cpf(rng: Rng): string {
  const n: number[] = [];
  for (let i = 0; i < 9; i += 1) n.push(rng.inteiro(0, 9));
  for (let k = 0; k < 2; k += 1) {
    let s = 0;
    for (let i = 0; i < n.length; i += 1) s += (n.length + 1 - i) * (n[i] as number);
    const d = (s * 10) % 11;
    n.push(d === 10 ? 0 : d);
  }
  const d = (i: number) => String(n[i]);
  return `${d(0)}${d(1)}${d(2)}.${d(3)}${d(4)}${d(5)}.${d(6)}${d(7)}${d(8)}-${d(9)}${d(10)}`;
}

function cep(rng: Rng): string {
  // Inclui prefixo com zero à esquerda de propósito: é a variação que quebra
  // implementação ingênua e que um exemplo escolhido a dedo esconde.
  const prefixo = String(rng.inteiro(1000, 99999)).padStart(5, "0");
  const sufixo = String(rng.inteiro(0, 999)).padStart(3, "0");
  return `${prefixo}-${sufixo}`;
}

function telefone(rng: Rng): string {
  const ddd = rng.escolher(DDDS);
  const a = String(rng.inteiro(1000, 9999));
  const b = String(rng.inteiro(1000, 9999));
  return `+${CODIGO_PAIS} ${ddd} 9${a}-${b}`;
}

/** Padrão Mercosul: LLLNLNN. */
function placa(rng: Rng): string {
  let letras = "";
  for (let i = 0; i < 3; i += 1) letras += rng.escolher([...LETRAS]);
  return `${letras}${rng.inteiro(0, 9)}${rng.escolher([...LETRAS])}${rng.inteiro(0, 9)}${rng.inteiro(0, 9)}`;
}

function email(rng: Rng): string {
  const palavra = (min: number, max: number) => {
    let s = "";
    const n = rng.inteiro(min, max);
    for (let i = 0; i < n; i += 1) s += rng.escolher([...MINUSCULAS]);
    return s;
  };
  const sep = rng.escolher([".", "_", ""]);
  return `${palavra(4, 9)}${sep}${palavra(4, 9)}@${rng.escolher(DOMINIOS)}`;
}

export type Pii = {
  cpf: string;
  cep: string;
  email: string;
  telefone: string;
  placa: string;
};

/** Semeado ⇒ determinístico. O seed é sempre explícito no teste. */
export function gerarPii(seed: number): Pii {
  const rng = lcg(seed);
  return {
    cpf: cpf(rng),
    cep: cep(rng),
    email: email(rng),
    telefone: telefone(rng),
    placa: placa(rng),
  };
}

/**
 * A fala e os valores, para o teste assertar sobre os dois.
 *
 * O formato imita o do dataset: PII solta em texto livre, misturada com o dado
 * de qualificação (a idade) que **não** é PII e não pode ser mascarado — cotar
 * depende dele.
 */
export function fraseDoLead(seed: number): [string, Pii] {
  const p = gerarPii(seed);
  return [
    `Tenho 35 anos, CPF ${p.cpf}, CEP ${p.cep}, ` +
      `meu email é ${p.email} e o whats é ${p.telefone}, ` +
      `placa ${p.placa}`,
    p,
  ];
}
